File: go/cpu/sort.go
package main

import (
	"cmp"
	"fmt"
	"slices"

	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/common"
	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/harness"
)

// sort-u64: the standard library's unstable sort of random uint64s
// (Go: pattern-defeating quicksort; Rust: ipnsort). Copying the unsorted
// input back before each iteration is untimed.
type sortU64 struct{ orig, work []uint64 }

func (w *sortU64) Prepare() { copy(w.work, w.orig) }

func (w *sortU64) Run() uint64 {
	slices.Sort(w.work)
	n := len(w.work)
	return w.work[0] ^ w.work[n/2] ^ w.work[n-1]
}

func (w *sortU64) Check() uint64 {
	d := common.NewDigest()
	for i, x := range w.work {
		if i > 0 && w.work[i-1] > x {
			panic(fmt.Sprintf("not sorted at %d", i))
		}
		d.Add(x)
	}
	return d.Sum()
}

func setupSortU64(p harness.Params) (harness.Instance, harness.Work, error) {
	n := int(p.Int("n", 1_000_000))
	r := common.NewSplitMix64(uint64(p.Int("seed", 1)))
	orig := make([]uint64, n)
	for i := range orig {
		orig[i] = r.Next()
	}
	return &sortU64{orig: orig, work: make([]uint64, n)}, harness.Work{Unit: "elements", PerRun: float64(n)}, nil
}

// sort-stable: stable sort of 16-byte records by a key with many duplicates
// (Go: insertion-sort blocks + in-place SymMerge; Rust: driftsort with an
// auxiliary buffer). Stability makes the output unique.
type rec struct{ key, val uint64 }

type sortStable struct{ orig, work []rec }

func (w *sortStable) Prepare() { copy(w.work, w.orig) }

func (w *sortStable) Run() uint64 {
	slices.SortStableFunc(w.work, func(a, b rec) int { return cmp.Compare(a.key, b.key) })
	n := len(w.work)
	return w.work[0].key ^ w.work[n/2].val ^ w.work[n-1].key
}

func (w *sortStable) Check() uint64 {
	d := common.NewDigest()
	for _, r := range w.work {
		d.Add(r.key)
		d.Add(r.val)
	}
	return d.Sum()
}

func setupSortStable(p harness.Params) (harness.Instance, harness.Work, error) {
	n := int(p.Int("n", 1_000_000))
	r := common.NewSplitMix64(uint64(p.Int("seed", 2)))
	kmax := uint64(max(n/4, 1))
	orig := make([]rec, n)
	for i := range orig {
		orig[i] = rec{key: r.Below(kmax), val: uint64(i)}
	}
	return &sortStable{orig: orig, work: make([]rec, n)}, harness.Work{Unit: "elements", PerRun: float64(n)}, nil
}

File: rust/cpu/src/sort.rs
//! Standard-library sorting.
//! * `sort-u64`: unstable sort of random u64 (Rust ipnsort vs Go pdqsort).
//! * `sort-stable`: stable sort of 16-byte records by a key with duplicates
//!   (Rust driftsort vs Go insertion blocks + in-place SymMerge).

use bench_common::{Digest, Instance, Params, SplitMix64, Work};

struct SortU64 {
    orig: Vec<u64>,
    work: Vec<u64>,
}

impl Instance for SortU64 {
    fn prepare(&mut self) {
        self.work.copy_from_slice(&self.orig);
    }
    fn run(&mut self) -> u64 {
        self.work.sort_unstable();
        let n = self.work.len();
        self.work[0] ^ self.work[n / 2] ^ self.work[n - 1]
    }
    fn has_check(&self) -> bool {
        true
    }
    fn check(&mut self) -> u64 {
        let mut d = Digest::new();
        for (i, &x) in self.work.iter().enumerate() {
            assert!(i == 0 || self.work[i - 1] <= x, "not sorted at {i}");
            d.add(x);
        }
        d.sum()
    }
}

pub fn setup_u64(p: &Params) -> Result<(Box<dyn Instance>, Work), String> {
    let n = p.int("n", 1_000_000) as usize;
    let mut r = SplitMix64::new(p.int("seed", 1) as u64);
    let orig: Vec<u64> = (0..n).map(|_| r.next_u64()).collect();
    Ok((
        Box::new(SortU64 { orig, work: vec![0; n] }),
        Work { unit: "elements", per_run: n as f64, input_bytes: 0 },
    ))
}

#[derive(Clone, Copy, Default)]
struct Rec {
    key: u64,
    val: u64,
}

struct SortStable {
    orig: Vec<Rec>,
    work: Vec<Rec>,
}

impl Instance for SortStable {
    fn prepare(&mut self) {
        self.work.copy_from_slice(&self.orig);
    }
    fn run(&mut self) -> u64 {
        self.work.sort_by_key(|r| r.key);
        let n = self.work.len();
        self.work[0].key ^ self.work[n / 2].val ^ self.work[n - 1].key
    }
    fn has_check(&self) -> bool {
        true
    }
    fn check(&mut self) -> u64 {
        let mut d = Digest::new();
        for r in &self.work {
            d.add(r.key);
            d.add(r.val);
        }
        d.sum()
    }
}

pub fn setup_stable(p: &Params) -> Result<(Box<dyn Instance>, Work), String> {
    let n = p.int("n", 1_000_000) as usize;
    let mut r = SplitMix64::new(p.int("seed", 2) as u64);
    let kmax = (n / 4).max(1) as u64;
    let orig: Vec<Rec> = (0..n).map(|i| Rec { key: r.below(kmax), val: i as u64 }).collect();
    Ok((
        Box::new(SortStable { orig, work: vec![Rec::default(); n] }),
        Work { unit: "elements", per_run: n as f64, input_bytes: 0 },
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testutil::{golden_cpu, params, run_once};
    use bench_common::golden_hex;

    #[test]
    fn matches_golden() {
        let g = golden_cpu();
        for (name, setup) in
            [("sort_u64", setup_u64 as bench_common::harness::SetupFn), ("sort_stable", setup_stable)]
        {
            let c = &g[name];
            let (mut inst, _) =
                setup(&params(&[("n", c["n"].to_string()), ("seed", c["seed"].to_string())])).unwrap();
            let (r, ch) = run_once(&mut inst);
            assert_eq!(r, golden_hex(&c["run"]), "{name}");
            assert_eq!(ch, golden_hex(&c["check"]), "{name}");
        }
    }
}
